Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -746,6 +783,35 @@ public class ProgramBuilder {
                     add(instr)
                 }
             }
+
+            // When we encounter a block begin:
+            // 1. We ensure that the context being opened removes atleast one required context
+            // 2. The required context is not empty
+            // 3. The required context is not equal to self.context (we don't want to capture blocks don't open atleast one requried context)
```

Comment:
Is this really true? Say the host program is inside a `BeginAsyncFunction` and the target program is
```
BeginFunction -> v7
BeginAsyncFunction
Await <- v7
```
So we'd now decide that we don't need the `BeginAsyncFunction`, then take the `BeginFunction`, and so end up with
```
BeginAsyncFunction       // From host
BeginFunction
Await
```
Which is wrong.

Reply:
Just pushed a couple of test for the scenario above: `testSameContextSplicing` and `testSameContextSplicing2`.

the algorithm will capture both `BeginAsync` and `BeginFunc` blocks. So the final program would be:

```
BeginAsync            // From the host program
 BeginAsync  v1         // From the target
  BeginFunction
   Await v1
  EndFunction
 EndAsync
EndAsync
```
So the reason the `BeginAsync` from the target is captured is due to the innerOutput `v1` that's required. 
In the scenario that we don't have a data dependency, then the algorithm would only capture `Await` and splice that into the host program.

Reply:
> In the scenario that we don't have a data dependency, then the algorithm would only capture Await and splice that into the host program.

Yeah that's the scenario I had in mind (and `Await` needs to have a data dependency on the `BeginFunction`). Don't you then end up with
```
BeginAsyncFunction       // From host
BeginFunction v1         // From target
Await v1
...
```
Which is invalid?

Reply:
Just so that I understand the scenario correctly, the host program would be:
```
BeginAsyncFunc
// point of splice
EndAsyncFunc
```
The target program would be:
```
BeginAsyncFunc
 BeginPlainFunc v1
  Await v1
 EndPlainFunc
EndAsyncFunc
```
and we start the splice at Await. Would that be right?
We wouldn't be able to generate the target program since `Await` cannot exist within a `PlainFunc` block. Starting a function block creates a new context and does not propagate the previous context.

Reply:
Not quite. Host program
```
BeginAsyncFunc
// point of splice
EndAsyncFunc
```

Target program:
```
BeginPlainFunc v1
  BeginAsyncFunc
    Await v1
  EndAsyncFunc
EndPlainFunc
```
and then start at the `Await`.

Now if we don't take the `BeginAsyncFunc` because `The required context is not equal to self.context` (from the comment), then we'd end up with

```
BeginAsyncFunc
  BeginPlainFunc v1
    Await v1
  EndPlainFunc
EndAsyncFunc
```

If I understand correctly, which would be incorrect.

Reply:
Right, so this scenario is currently captured in testcase: `testSameContextSplicing2` which is under ProgramBuilderTests.
So the splice in your example above would capture the entire target program and splice that into the host. So we end up with: 
```
BeginAsync            // From the host program
 BeginPlain  v1         // From the target
  BeginAsyncFunction // From the target
   Await v1
  EndAsyncFunction
 EndPlain
EndAsync
```
I think the comment I've added is misleading and that's probably the source of confusion. An example that demonstrates why this check is necessary can be found in testcase: `testSplicing3`. Without the check we end up capturing the `BeginFunc` and `EndFunc` block which isn't required for the host program since neither does it have data dependencies nor does it open a required context that doesn't already exist in the host program. Also, now that I've looked at it again, `testSplicing3` is the only testcase that would fail without this check.

Reply:
Ah great, yeah that test does indeed cover the scenario!
I think the reason why what I'm describing doesn't happen is because when the splicing loop reaches the `BeginAsyncFunction` from the target, the `requiredContext = [.async]`, but the `self.context = [.script, .async]`, and so `requiredContext != self.context`. This is because 
```swift
class Await: Operation {
    init() {
        super.init(numInputs: 1, numOutputs: 1, attributes: [], requiredContext: .asyncFunction)
    }
}
```
so probably, here `requiredContext` should really be `[.asyncFunction, .script]`, since the other operations also have `.script` as their `requiredContext` (since it's the default value if none is provided)?
From what I can tell, `testSplicing3` doesn't include an instruction that requires `[.function]` context, and so doesn't need the `BeginFunctionDefinition`.

Reply:
This problem should now be resolved with adding the intersection check: 
```
 if requiredContext.subtracting(instr.op.contextOpened) != requiredContext &&
    requiredContext.intersection(instr.op.contextOpened) != .script && 
    requiredContext != .empty {
                    requiredContextStack.append(requiredContext)
                    add(instr)
                    requiredContext = requiredContextStack.removeLast()
} 
```
We make sure that the default context isn't the only one that's being removed from `requiredContexts`.